When using the workaround for the Lint issue, Spoon does not run with multiple devices anymore

When using the workaround fixes the lint issue (described on the main page of spoon-gradle-plugin), Spoon does not work anymore with multiple devices are attached. One device still works though.

This output is from a run of 2 tests on 2 devices:
```
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR Exception in thread "main" java.lang.NoClassDefFoundError: com/google/common/collect/Multimap
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR Exception in thread "main" java.lang.NoClassDefFoundError: com/google/common/collect/Multimap
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.lang.Class.getDeclaredMethods0(Native Method)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.lang.Class.getDeclaredMethods0(Native Method)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.lang.Class.privateGetDeclaredMethods(Class.java:2693)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.lang.Class.privateGetDeclaredMethods(Class.java:2693)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.lang.Class.privateGetMethodRecursive(Class.java:3040)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.lang.Class.privateGetMethodRecursive(Class.java:3040)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.lang.Class.getMethod0(Class.java:3010)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.lang.Class.getMethod0(Class.java:3010)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.lang.Class.getMethod(Class.java:1776)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.lang.Class.getMethod(Class.java:1776)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:544)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:544)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:526)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:526)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR Caused by: java.lang.ClassNotFoundException: com.google.common.collect.Multimap
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR Caused by: java.lang.ClassNotFoundException: com.google.common.collect.Multimap
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.net.URLClassLoader$1.run(URLClassLoader.java:372)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.net.URLClassLoader$1.run(URLClassLoader.java:372)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.net.URLClassLoader$1.run(URLClassLoader.java:361)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.security.AccessController.doPrivileged(Native Method)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.security.AccessController.doPrivileged(Native Method)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.net.URLClassLoader.findClass(URLClassLoader.java:360)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:308)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
2015-01-20 17:21:16 [SDR.printStream] [4df1cb5a5ef88fa7] STDERR ... 7 more
2015-01-20 17:21:16 [SDR.printStream] [034b2915093ab9bf] STDERR ... 7 more
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -9,7 +9,7 @@ install:
   - TERM=dumb ./gradlew -q assemble
 
 script:
-  - TERM=dumb ./gradlew check
+  - TERM=dumb ./gradlew integrationTest
 
 after_success:
   - ./scripts/deploy_to_maven.sh

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,15 +1,33 @@
-language: java
+language: android
+
+android:
+  components:
+    - build-tools-22.0.0
+    - android-21
+    - sys-img-armeabi-v7a-android-21
+    - extra-android-m2repository
+  licenses:
+    - 'android-sdk-license-.+'
 
 env:
   global:
     - secure: "lPttF+AikE2/Rtw5CO49XMMD2cLEdOUnswmtMX5rIg4Ot8XZ3zx0tDan/8bvAENZZh2xLbD2e+sHrCCSVpmyH+nHBRu5fMzUZww30gV8GDgwWJqKmQT/TUUdrpPFEYDfEFhy0WtDnw7LO/DwB3MjQVMrp9huzm8UYBGwSIiQ6GI="
     - secure: "Z4LXYWoxV2ZAL3onkc5YIWUo1z8ftVSS6nvfEpc6pBWam5HbTM7XVwzbyr0d9edT/1H+/rVyyZQJ9j1JZjpvX9WA+0RVMTgYU2aVdV/EyaQesNYDZfBr5CUteEj3GFxEIzy0wgWgPfbUb44K577LyB6ra9TdZUdEngCTbXCUmwo="
 
+before_install:
+  - export TERM=dumb
+
+  # Create and run emulator
+#  - echo no | android create avd --force -n test -t android-21 --abi armeabi-v7a
+#  - emulator -avd test -no-skin -no-audio -no-window &
+
 install:
-  - TERM=dumb ./gradlew -q assemble
+  - ./gradlew -q assemble
 
 script:
-  - TERM=dumb ./gradlew integrationTest
+#  - android-wait-for-emulator || exit 1
+#  - adb shell input keyevent 82 &
+  - ./gradlew integrationTest
 
 after_success:
   - ./scripts/deploy_to_maven.sh

--- a/README.md
+++ b/README.md
@@ -17,7 +17,7 @@ buildscript {
   }
 
   dependencies {
-    classpath 'com.stanfy.spoon:spoon-gradle-plugin:0.14.1'
+    classpath 'com.stanfy.spoon:spoon-gradle-plugin:1.0.0'
   }
 }
 
@@ -91,7 +91,7 @@ If you have troubles with running Android Lint after this plugin is applied, try
 buildscript {
   // ...
   dependencies {
-    classpath('com.stanfy.spoon:spoon-gradle-plugin:0.14.1') {
+    classpath('com.stanfy.spoon:spoon-gradle-plugin:1.0.0') {
       exclude module: 'guava'
     }
     classpath 'com.google.guava:guava:17.0'
@@ -100,6 +100,13 @@ buildscript {
 ```
 You can find details in the [related issue](https://github.com/stanfy/spoon-gradle-plugin/issues/33).
 
+Testing
+-------
+After making editions, run
+```
+./gradlew integrationTest
+```
+This will install the plugin to local maven repo and run spoon task in the example project.
 
 License
 -------

--- a/build.gradle
+++ b/build.gradle
@@ -12,7 +12,7 @@ apply plugin: 'nexus'
 apply plugin: 'project-report'
 
 group = 'com.stanfy.spoon'
-version = '0.14.2-SNAPSHOT'
+version = '1.0.0-SNAPSHOT'
 
 jar {
   manifest {
@@ -79,5 +79,5 @@ modifyPom {
 
 task integrationTest(type: GradleBuild, dependsOn: ['check', 'install']) {
   dir = file('example')
-  tasks = ['check']
+  tasks = ['clean', 'spoon']
 }

--- a/example/build.gradle
+++ b/example/build.gradle
@@ -6,6 +6,6 @@ buildscript {
 
   dependencies {
     classpath 'com.android.tools.build:gradle:1.1.3'
-    classpath 'com.stanfy.spoon:spoon-gradle-plugin:0.14.2-SNAPSHOT'
+    classpath 'com.stanfy.spoon:spoon-gradle-plugin:1.0.0-SNAPSHOT'
   }
 }

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -3,4 +3,4 @@ distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=http\://services.gradle.org/distributions/gradle-2.2.1-all.zip
+distributionUrl=http\://services.gradle.org/distributions/gradle-2.3-all.zip

--- a/src/main/groovy/com/stanfy/spoon/gradle/SpoonRunTask.groovy
+++ b/src/main/groovy/com/stanfy/spoon/gradle/SpoonRunTask.groovy
@@ -6,6 +6,7 @@ import groovy.transform.PackageScope
 import org.gradle.api.DefaultTask
 import org.gradle.api.GradleException
 import org.gradle.api.Project
+import org.gradle.api.artifacts.ResolvedDependency
 import org.gradle.api.tasks.*
 import org.slf4j.Logger
 import org.slf4j.LoggerFactory
@@ -144,38 +145,75 @@ class SpoonRunTask extends DefaultTask implements VerificationTask {
     }
   }
 
-  private def findPluginDependency(final Project project) {
-    def pluginDep = null
-    //noinspection GroovyAssignabilityCheck
-    project.buildscript.configurations.each {
-      if (pluginDep) {
-        return
-      }
+  private static def findCpDependency(final Project project, final String prefix) {
+    def configuration = project.buildscript.configurations.classpath
+    return configuration.resolvedConfiguration.firstLevelModuleDependencies.find {
+      it.name.startsWith prefix
+    }
+  }
 
-      pluginDep = it.resolvedConfiguration.firstLevelModuleDependencies.find {
-        it.name.startsWith SpoonRunTask.PLUGIN_DEP_NAME
+  private def lookupDependency(final String prefix) {
+    def dep = null
+    def p = project, pp = null
+    while (!dep && p) {
+      pp = p
+      dep = findCpDependency(p, prefix)
+      p = p.parent
+    }
+    return [dep, pp]
+  }
+
+  private static void collectDependencies(ResolvedDependency root, def classpath, def allDeps) {
+    root.children.each { dep ->
+      def conflict = allDeps.find { it.moduleName == dep.moduleName && it.moduleGroup == dep.moduleGroup }
+      if (conflict && conflict.moduleVersion != dep.moduleVersion) {
+        LOG.warn("There is a dependencies conflict for ${dep.moduleGroup}:${dep.modeuleName}. "
+            + "Versions: ${conflict.modeuleVersion} and ${dep.modeuleVersion}")
+      } else {
+        allDeps.add dep
       }
     }
-    return pluginDep
+    classpath.addAll root.allModuleArtifacts.collect { artifact -> artifact.file }
+    root.children.each {
+      collectDependencies it, classpath, allDeps
+    }
+  }
+
+  private static void checkDependencies(ResolvedDependency root, def classpath, def allDeps) {
+    root.children.each { dep ->
+      def existing = allDeps.find { it.moduleName == dep.moduleName && it.moduleGroup == dep.moduleGroup }
+      if (existing) {
+        collectDependencies(dep, classpath, allDeps)
+      }
+    }
+    root.children.each { dep ->
+      checkDependencies(dep, classpath, allDeps)
+    }
   }
 
   private String getClasspath() {
-    def pluginDep = null
-    def p = project
-    while (!pluginDep && p) {
-      pluginDep = findPluginDependency(p)
-      p = p.parent
-    }
+    def (pluginDep, usedProject) = lookupDependency(PLUGIN_DEP_NAME)
     if (!pluginDep) {
       throw new IllegalStateException("Could not resolve spoon dependencies")
     }
 
-    def spoon = pluginDep.children.find { it.name.startsWith SpoonRunTask.SPOON_DEP_NAME }
+    ResolvedDependency spoon = pluginDep.children.find { it.name.startsWith SpoonRunTask.SPOON_DEP_NAME } as ResolvedDependency
     if (!spoon) { throw new IllegalStateException("Cannot find spoon-runner in dependencies") }
     // Collect spoon dependencies.
-    def classpath = spoon.allModuleArtifacts.collect { it.file }
+    def classpath = new HashSet(), allDeps = []
+    collectDependencies(spoon, classpath, allDeps)
     // Add spoon runner.
     classpath += pluginDep.allModuleArtifacts.find { it.name == SpoonRunTask.SPOON_RUNNER_ARTIFACT }.file
+
+    /*
+     * XXX Due to how Gradle dependencies resolution works we can get not all the required deps from the step above.
+     *     For example spoon dep is com.android.tools.ddms:ddmlib:23.2.1.
+     *     If you use Android plugin 1.1.3, ddmlib will be resolved to 24.1.3. And its dependencies will not traversed yet.
+     *     Hence, we traverse the whole tree again and pick up those parts of the tree that we miss.
+     */
+    usedProject.buildscript.configurations.classpath.resolvedConfiguration.firstLevelModuleDependencies.each {
+      SpoonRunTask.checkDependencies((ResolvedDependency) it, classpath, allDeps)
+    }
 
     return project.files(classpath).asPath
   }

